aspnetcore/blazor/Security-guidance.md

```diff
@@ -0,0 +1,254 @@
+* It will timeout any .NET to JS call after 1 minute without receiving a response from the client.
+* It performs some basic validation in all input comming from the browser from JS interop.
+  * Ensure that the dotnet references are valid and of the type expected by the .NET method.
+  * Ensure that the data is not malformed.
+  * Ensure the right number of arguments for the method is present on the payload.
+  * Ensure that the arguments/result can be deserialized correctly before invoking the method.
+* It performs some basic validation in all input comming from the browser from dispatched events.
+  * It validates that the event has a valid type.
+  * It validates that the data for the event can be deserialized.
+  * It validates that there is an event handler associated with that event.
+
+For a XSS vulnerability to be possible, the application must allow some user input to be included as part of the rendered page. Blazor server-side components go through a compile-time step where the contents of the page are transformed into what we call a RenderTree (a virtual DOM like structure) and the diffs between previous and current renders are sent to the client for rendering, or in the case of prerendering serialized into HTML.
+* At this step blazor server-side takes care of HTML encoding all content that wasn't defined as static markup at compile-time.
+* Script tags aren't allowed to be included as components and a compile time error is generated when one such tag is included in a razor page.
+  * Script tags will not behave the way one would expect on the page. Once added they execute the contents of the script and those side-effects don't disappear even if you remove the script tag.
+* Component authors can author components in C# without using Razor. For these cases, the component author is responsible for encoding all potentiall user input from the component.
+* It is also recommended that you avoid including script tags as part of your component render tree for the reasons stated above.
+
+As part of protecting against XSS attacks we recommend looking into application specific mitigations like content security policy (CSP). See https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP for more details.
+
+## Cross-origin protection
+Cross-origin attacks involve a client from a different origin performing some action against your server. This action is typically a form POST (CSRF), but opening a websocket is also possible. Blazor server-side applications offer the same guarantees that any other SignalR application using the hub protocol offers.
+* A browser from a different origin can't connect to the blazor-server side application and start a new session (circuit) with the default application configuration.
+  * This is due to CORS being disabled by default and SignalR forcing a preflight request against the server as part of the hub protocol handshake.
```

Comment:
> SignalR forcing a preflight request

Just for my info, could you recap how this is achieved? Is there any way a malicious client (with a modified version of `signalr-client.js`) could skip the HTTP request that precedes the websocket connection request? For example, could it use server-side code to get whatever information is needed from the initial HTTP request, send that data to the client, and then directly open a websocket connection from there?

Reply:
The server checks for X-Requested-With afaik and fails if not. Is that not the case?

Reply:
You are right, the server doesn't enforce anything. The client sends the header, but the server does nothing.

Reply:
@BrennanConroy You are right. We need to change the guidance here to point to the way to limit connections from different origins in SignalR. A legitimate client won't get through to the app due to the x-http-requested-with , but an unofficial client might

@BrennanConroy @anurse Is this something we could add to the HubProtocol? (Allowing to configure the negotiate call on a per hub basis to require the presence of x-requested-with)